## Interview questions 

https://drive.google.com/drive/folders/1BrZ3ZRDYvwGiiFoH13PEGv-c9RQ0-vWE

https://colab.research.google.com/drive/1zfOkg1XCkXVHBQ26soQgPFL3eVFhHLfS - Linear regression notebook

## Linear regression



### Imports

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import pickle
from pandas_profiling import ProfileReport
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge, Lasso, RidgeCV, LassoCV, ElasticNet, ElasticNetCV
from sklearn.model_selection import train_test_split
import statsmodels.api as sm
import seaborn as sns

In [2]:
df = pd.read_csv('Advertising.csv')
df

Unnamed: 0     TV  Radio  Newspaper  Sales
0             1  230.1   37.8       69.2   22.1
1             2   44.5   39.3       45.1   10.4
2             3   17.2   45.9       69.3    9.3
3             4  151.5   41.3       58.5   18.5
4             5  180.8   10.8       58.4   12.9
..          ...    ...    ...        ...    ...
195         196   38.2    3.7       13.8    7.6
196         197   94.2    4.9        8.1    9.7
197         198  177.0    9.3        6.4   12.8
198         199  283.6   42.0       66.2   25.5
199         200  232.1    8.6        8.7   13.4

[200 rows x 5 columns]

### Defining the problem 
    > We have 3 columns regarding 3 various types of advertisement:
        * TV,
        * Radio, 
        * Newspaper.

Our task is to study, how much the differnt types of advertisement impact the sales. And build a model, which will predict the impact on sales.  

Task #2 - Were the amount of money invest in the specific type of adverisement will allow to achieve the goal in sales?
Task #3 - If there are some relations between variables? If they are we can try to implement Linear Regression.

In [3]:
# Basic EDA
df.head()

Unnamed: 0     TV  Radio  Newspaper  Sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [4]:
df.describe()

Unnamed: 0          TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000  200.000000
mean   100.500000  147.042500   23.264000   30.554000   14.022500
std     57.879185   85.854236   14.846809   21.778621    5.217457
min      1.000000    0.700000    0.000000    0.300000    1.600000
25%     50.750000   74.375000    9.975000   12.750000   10.375000
50%    100.500000  149.750000   22.900000   25.750000   12.900000
75%    150.250000  218.825000   36.525000   45.100000   17.400000
max    200.000000  296.400000   49.600000  114.000000   27.000000

In [5]:
# Pandas profiling
ProfileReport(df)

In [6]:
# Saving PF to variable for beautification
pf = ProfileReport(df)
pf.to_widgets()

In [7]:
# Saving report
pf.to_file('pf_advertising.html')

### Multicolinearity 

Two independent variable which are highly correlated to each other. In this case one of them needs to be dropped. Why? Because upon feeding the data to the model, it will start to learn correlation in the dataset (x1, x2) and not work on the correlations between x and y. 


> In this dataset we do not have multicolinearity. We do not need to drop any columns, because of it. 

## Building a model

Building Linear regression model with Tv and Sales. (the highest correlation)

y = mx + c

y - predictions
m - 
c - 

In [8]:
# Selecting variables X, y
X = df[['TV']]
y = df['Sales']

### Using sklearn

In [9]:
from sklearn.linear_model import LinearRegression

In [10]:
lin_reg = LinearRegression()

In [11]:
lin_reg.fit(X, y)

LinearRegression()

#### Checking the values of m and c

In [12]:
lin_reg.intercept_ # c value

7.032593549127693

In [13]:
lin_reg.coef_ # m value

array([0.04753664])

#### Equation for sales

    > sales = m*x + c

In [14]:
# Saving model to file
model = 'lin_reg.sav'
pickle.dump(lin_reg, open(model,'wb'))

#### Predictions

In [15]:
# Predicting 45 on Tv, results in sales
lin_reg.predict([[45]])

array([9.17174237])

In [16]:
# Predicting on multiple values
l = [4,5,6,7,89,34,45,67,23]

for i in l:
    print(lin_reg.predict([[i]]))

[7.22274011]
[7.27027675]
[7.31781339]
[7.36535003]
[11.26335455]
[8.64883932]
[9.17174237]
[10.21754846]
[8.12593628]


In [17]:
# Loading saved model
sav_model = pickle.load(open(model, 'rb'))

In [18]:
# Preciting with saved, loaded model
sav_model.predict([[45]])

array([9.17174237])

#### Evaluations of the model

Note that evaluation on the same data as we are doing training is done here only for learning purposes. We shouldn't do evaluation on the same data. 

In [19]:
lin_reg.score(X,y) # Evalution with rsquared 

0.611875050850071

## Multilinear Regression

In [20]:
data_x = df[['TV', 'Radio', 'Newspaper']]

In [21]:
data_x

TV  Radio  Newspaper
0    230.1   37.8       69.2
1     44.5   39.3       45.1
2     17.2   45.9       69.3
3    151.5   41.3       58.5
4    180.8   10.8       58.4
..     ...    ...        ...
195   38.2    3.7       13.8
196   94.2    4.9        8.1
197  177.0    9.3        6.4
198  283.6   42.0       66.2
199  232.1    8.6        8.7

[200 rows x 3 columns]

In [22]:
y = df.Sales

In [23]:
lin_reg.fit(data_x, y)

LinearRegression()

In [24]:
lin_reg.intercept_, lin_reg.coef_

(2.9388893694594085, array([ 0.04576465,  0.18853002, -0.00103749]))

### Equation for multilinear regression

y = mx1 + mx2 + mx3 + c
y = 0.04 (TV) + 0.188 (Radio) - 0.00103 (Newspaper) + 2.938 (intercept -> c)


### What will be the Confidence Interval on r-squared?

In [25]:
lin_reg.score(data_x, y)

0.8972106381789522

**Conclusion**
    > Score is better for multilinear regression. Multiple columns output better predictions. 

#### Checking r-squared for TV and radio

In [26]:
sel_data = df[['TV', 'Radio']]

In [27]:
lin_reg.fit(sel_data, y)

LinearRegression()

In [28]:
    lin_reg.score(sel_data, y)

0.8971942610828956

**Conlusion**
    > From the data we can see there is no difference in score between having 2 or 3 columns. It seems that investment in newspaper advertisement is not changing anything to the score in this case. 
    
    > Base on the results, we can conclude, that Newspaper column should be dropped. It is not affecting the sales in positive direction.  
    
Note! That r-squared is not only way to evaluate features scores. 

### Evaluation base on statistical models - Feature selection

In [29]:
import statsmodels.formula.api as smf

In [30]:
lm = smf.ols(formula = 'Sales ~ TV', data=df).fit() # Ordinary Least Squares
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.612
Model:                            OLS   Adj. R-squared:                  0.610
Method:                 Least Squares   F-statistic:                     312.1
Date:                Thu, 10 Mar 2022   Prob (F-statistic):           1.47e-42
Time:                        13:50:37   Log-Likelihood:                -519.05
No. Observations:                 200   AIC:                             1042.
Df Residuals:                     198   BIC:                             1049.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0326      0.458     15.360      0.000       6.130       7.935
TV             0.0475      0.003     17.668      0.000       0.042       0.053
==============================================================================
Omnibus:                        0.531   Durbin-Watson:                   1.935
Prob(Omnibus):                  0.767   Jarque-Bera (JB):                0.669
Skew:                          -0.089   Prob(JB):                        0.716
Kurtosis:                       2.779   Cond. No.                         338.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [31]:
lm = smf.ols(formula = 'Sales ~ TV + Radio', data=df).fit() # Ordinary Least Squares
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     859.6
Date:                Thu, 10 Mar 2022   Prob (F-statistic):           4.83e-98
Time:                        13:50:37   Log-Likelihood:                -386.20
No. Observations:                 200   AIC:                             778.4
Df Residuals:                     197   BIC:                             788.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9211      0.294      9.919      0.000       2.340       3.502
TV             0.0458      0.001     32.909      0.000       0.043       0.048
Radio          0.1880      0.008     23.382      0.000       0.172       0.204
==============================================================================
Omnibus:                       60.022   Durbin-Watson:                   2.081
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              148.679
Skew:                          -1.323   Prob(JB):                     5.19e-33
Kurtosis:                       6.292   Cond. No.                         425.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [32]:
lm = smf.ols(formula = 'Sales ~ TV + Radio + Newspaper', data=df).fit() # Ordinary Least Squares
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Thu, 10 Mar 2022   Prob (F-statistic):           1.58e-96
Time:                        13:50:37   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

#### Legend
* std err -> Standard deviation
* Adj R-squared -> Tells us about possible overfitting. It is more robust than the r-squared. It is considering in calculation(above the usual r-squared equation) increased increment of the rows and a number of a columns. 

> Take a look at p-value of Newspaper. It has .860 value, which means it has very low sigificance to the model.

What is the meaning of p-value of the Newspaper here? In 100 experiments, only 14 are contributing to the final score. 86 of them are not contributing. 

**Conclusion**
    > Taking into consideration both r-squared and p-value of the Newspaper, we can drop Newspaper. 

### Multicolinearity - variance inflation factor - Feature selection

To check for Multicolinearity calculate variance inflation factor. 

### Regularization

Adjusting values of m and c to achieve the best results. 
Regularization is used to stabilize the outcome predictions. 

#### L1 - REGULARIZATION - Lasso Regression
> Having higher impact. - for smaller values of m


#### L2 - Ridge Regression
> Having less impact. - for smaller values of m

#### Elastic net
> L1, L2 combined. It is changing behaviour according to the need. 

## Loading new data

In [33]:
pd.read_csv('Admission_Prediction.csv')

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0             1      337.0        118.0                4.0  4.5  4.5  9.65   
1             2      324.0        107.0                4.0  4.0  4.5  8.87   
2             3        NaN        104.0                3.0  3.0  3.5  8.00   
3             4      322.0        110.0                3.0  3.5  2.5  8.67   
4             5      314.0        103.0                2.0  2.0  3.0  8.21   
..          ...        ...          ...                ...  ...  ...   ...   
495         496      332.0        108.0                5.0  4.5  4.0  9.02   
496         497      337.0        117.0                5.0  5.0  5.0  9.87   
497         498      330.0        120.0                5.0  4.5  5.0  9.56   
498         499      312.0        103.0                4.0  4.0  5.0  8.43   
499         500      327.0        113.0                4.0  4.5  4.5  9.04   

     Research  Chance of Admit  
0           1             0.92  
1           1             0.76  
2           1             0.72  
3           1             0.80  
4           0             0.65  
..        ...              ...  
495         1             0.87  
496         1             0.96  
497         1             0.93  
498         0             0.73  
499         0             0.84  

[500 rows x 9 columns]

### Task
Predicting chance of admission based on given information. 

### EDA

In [34]:
# Storing data in variable adm
adm = pd.read_csv('Admission_Prediction.csv')

In [35]:
# Pandas profiling
ProfileReport(adm)

### Missing values

In [36]:
adm['GRE_Score'] = adm['GRE_Score'].fillna(adm['GRE_Score'].mean())

In [37]:
adm['TOEFL_Score'] = adm['TOEFL_Score'].fillna(adm['TOEFL_Score'].mean())

In [38]:
adm['University_Rating'] = adm['University_Rating'].fillna(adm['University_Rating'].mean())

In [40]:
adm.describe()

Serial_No.   GRE_Score  TOEFL_Score  University_Rating         SOP  \
count  500.000000  500.000000   500.000000         500.000000  500.000000   
mean   250.500000  316.558763   107.187755           3.121649    3.374000   
std    144.481833   11.103952     6.051338           1.128802    0.991004   
min      1.000000  290.000000    92.000000           1.000000    1.000000   
25%    125.750000  309.000000   103.000000           2.000000    2.500000   
50%    250.500000  316.558763   107.000000           3.000000    3.500000   
75%    375.250000  324.000000   112.000000           4.000000    4.000000   
max    500.000000  340.000000   120.000000           5.000000    5.000000   

             LOR        CGPA    Research  Chance_of_Admit  
count  500.00000  500.000000  500.000000        500.00000  
mean     3.48400    8.576440    0.560000          0.72174  
std      0.92545    0.604813    0.496884          0.14114  
min      1.00000    6.800000    0.000000          0.34000  
25%      3.00000    8.127500    0.000000          0.63000  
50%      3.50000    8.560000    1.000000          0.72000  
75%      4.00000    9.040000    1.000000          0.82000  
max      5.00000    9.920000    1.000000          0.97000

In [41]:
adm.isnull().sum()

Serial_No.           0
GRE_Score            0
TOEFL_Score          0
University_Rating    0
SOP                  0
LOR                  0
CGPA                 0
Research             0
Chance_of_Admit      0
dtype: int64

### Dropping id column

In [45]:
adm.drop('Serial_No.', axis=1, inplace=True)

In [46]:
adm

GRE_Score  TOEFL_Score  University_Rating  SOP  LOR  CGPA  Research  \
0    337.000000        118.0                4.0  4.5  4.5  9.65         1   
1    324.000000        107.0                4.0  4.0  4.5  8.87         1   
2    316.558763        104.0                3.0  3.0  3.5  8.00         1   
3    322.000000        110.0                3.0  3.5  2.5  8.67         1   
4    314.000000        103.0                2.0  2.0  3.0  8.21         0   
..          ...          ...                ...  ...  ...   ...       ...   
495  332.000000        108.0                5.0  4.5  4.0  9.02         1   
496  337.000000        117.0                5.0  5.0  5.0  9.87         1   
497  330.000000        120.0                5.0  4.5  5.0  9.56         1   
498  312.000000        103.0                4.0  4.0  5.0  8.43         0   
499  327.000000        113.0                4.0  4.5  4.5  9.04         0   

     Chance_of_Admit  
0               0.92  
1               0.76  
2               0.72  
3               0.80  
4               0.65  
..               ...  
495             0.87  
496             0.96  
497             0.93  
498             0.73  
499             0.84  

[500 rows x 8 columns]

### Splitting for X and y

In [48]:
y = adm['Chance_of_Admit']

In [63]:
x = adm

In [64]:
x

GRE_Score  TOEFL_Score  University_Rating  SOP  LOR  CGPA  Research
0    337.000000        118.0                4.0  4.5  4.5  9.65         1
1    324.000000        107.0                4.0  4.0  4.5  8.87         1
2    316.558763        104.0                3.0  3.0  3.5  8.00         1
3    322.000000        110.0                3.0  3.5  2.5  8.67         1
4    314.000000        103.0                2.0  2.0  3.0  8.21         0
..          ...          ...                ...  ...  ...   ...       ...
495  332.000000        108.0                5.0  4.5  4.0  9.02         1
496  337.000000        117.0                5.0  5.0  5.0  9.87         1
497  330.000000        120.0                5.0  4.5  5.0  9.56         1
498  312.000000        103.0                4.0  4.0  5.0  8.43         0
499  327.000000        113.0                4.0  4.5  4.5  9.04         0

[500 rows x 7 columns]

### Standarization of the data

In [65]:
# Creating object for Standard Scaler
scaler = StandardScaler()

In [66]:
# Fitting and transforming x
arr = scaler.fit_transform(x)

In [71]:
# Transforming to a data frame
df1 = pd.DataFrame(arr, columns=x.columns)

In [72]:
df1

GRE_Score  TOEFL_Score  University_Rating       SOP       LOR  \
0    1.842741e+00     1.788542           0.778906  1.137360  1.098944   
1    6.708143e-01    -0.031058           0.778906  0.632315  1.098944   
2    5.124333e-15    -0.527313          -0.107877 -0.377773  0.017306   
3    4.905178e-01     0.465197          -0.107877  0.127271 -1.064332   
4   -2.306679e-01    -0.692731          -0.994659 -1.387862 -0.523513   
..            ...          ...                ...       ...       ...   
495  1.392000e+00     0.134360           1.665688  1.137360  0.558125   
496  1.842741e+00     1.623124           1.665688  1.642404  1.639763   
497  1.211704e+00     2.119379           1.665688  1.137360  1.639763   
498 -4.109644e-01    -0.692731           0.778906  0.632315  1.639763   
499  9.412590e-01     0.961451           0.778906  1.137360  1.098944   

         CGPA  Research  
0    1.776806  0.886405  
1    0.485859  0.886405  
2   -0.954043  0.886405  
3    0.154847  0.886405  
4   -0.606480 -1.128152  
..        ...       ...  
495  0.734118  0.886405  
496  2.140919  0.886405  
497  1.627851  0.886405  
498 -0.242367 -1.128152  
499  0.767220 -1.128152  

[500 rows x 7 columns]

In [73]:
ProfileReport(df1)

In [75]:
df1.describe()

GRE_Score   TOEFL_Score  University_Rating           SOP  \
count  5.000000e+02  5.000000e+02       5.000000e+02  5.000000e+02   
mean   4.350520e-15  9.419132e-16       5.608847e-16  2.926548e-16   
std    1.001002e+00  1.001002e+00       1.001002e+00  1.001002e+00   
min   -2.394225e+00 -2.512331e+00      -1.881441e+00 -2.397950e+00   
25%   -6.814090e-01 -6.927310e-01      -9.946589e-01 -8.828175e-01   
50%    5.124333e-15 -3.105811e-02      -1.078766e-01  1.272712e-01   
75%    6.708143e-01  7.960330e-01       7.789057e-01  6.323155e-01   
max    2.113186e+00  2.119379e+00       1.665688e+00  1.642404e+00   

                LOR          CGPA      Research  
count  5.000000e+02  5.000000e+02  5.000000e+02  
mean  -1.332268e-17  3.091971e-15 -2.202682e-16  
std    1.001002e+00  1.001002e+00  1.001002e+00  
min   -2.686789e+00 -2.940115e+00 -1.128152e+00  
25%   -5.235128e-01 -7.430227e-01 -1.128152e+00  
50%    1.730621e-02 -2.720919e-02  8.864053e-01  
75%    5.581253e-01  7.672196e-01  8.864053e-01  
max    1.639763e+00  2.223672e+00  8.864053e-01

### Multicolinearity

In [76]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [77]:
arr.shape

(500, 7)

In [78]:
[variance_inflation_factor(arr,i) for i in range(arr.shape[1])] # Variance inflation factor for every columns

[4.153267572225826,
 3.792866110594647,
 2.508768242278762,
 2.7757495092534947,
 2.0373076624897517,
 4.651669561154733,
 1.4593106786827286]

In [79]:
vif_df = pd.DataFrame()

In [82]:
vif_df['vif'] = [variance_inflation_factor(arr,i) for i in range(arr.shape[1])]
vif_df['feature'] = x.columns

In [83]:
vif_df

feature       vif
0          GRE_Score  4.153268
1        TOEFL_Score  3.792866
2  University_Rating  2.508768
3                SOP  2.775750
4                LOR  2.037308
5               CGPA  4.651670
6           Research  1.459311

No need to remove any columns. Vif < 10. 

### Modelling
### Splitting - train_test split

In [85]:
x_train, x_test, y_train, y_test = train_test_split(arr, y, test_size=0.25, random_state=42)

In [86]:
lr = LinearRegression()

In [87]:
lr.fit(x_train, y_train)

LinearRegression()

In [89]:
# Saving model
pickle.dump(lr, open('admission_lr_model.pickle', 'wb'))

In [93]:
import os
!ls

'ls' is not recognized as an internal or external command,
operable program or batch file.


In [97]:
# Predictions
adm = pd.read_csv('Admission_Prediction.csv')
adm

Serial No.  GRE Score  TOEFL Score  University Rating  SOP  LOR  CGPA  \
0             1      337.0        118.0                4.0  4.5  4.5  9.65   
1             2      324.0        107.0                4.0  4.0  4.5  8.87   
2             3        NaN        104.0                3.0  3.0  3.5  8.00   
3             4      322.0        110.0                3.0  3.5  2.5  8.67   
4             5      314.0        103.0                2.0  2.0  3.0  8.21   
..          ...        ...          ...                ...  ...  ...   ...   
495         496      332.0        108.0                5.0  4.5  4.0  9.02   
496         497      337.0        117.0                5.0  5.0  5.0  9.87   
497         498      330.0        120.0                5.0  4.5  5.0  9.56   
498         499      312.0        103.0                4.0  4.0  5.0  8.43   
499         500      327.0        113.0                4.0  4.5  4.5  9.04   

     Research  Chance of Admit  
0           1             0.92  
1           1             0.76  
2           1             0.72  
3           1             0.80  
4           0             0.65  
..        ...              ...  
495         1             0.87  
496         1             0.96  
497         1             0.93  
498         0             0.73  
499         0             0.84  

[500 rows x 9 columns]

In [98]:
lr.predict([[337.000000,118.0,4.0,4.5,4.5,9.65,1]])

array([11.32582343])

In [99]:
# Predicting with scaled data
scaler.transform([[337.000000,118.0,4.0,4.5,4.5,9.65,1]])

array([[1.84274116, 1.78854223, 0.77890565, 1.13735981, 1.09894429,
        1.77680627, 0.88640526]])

In [100]:
lr.predict([[1.84274116, 1.78854223, 0.77890565, 1.13735981, 1.09894429,
        1.77680627, 0.88640526]])

array([0.95976005])

In [102]:
# Predicting with model loaded with pickle
model = pickle.load(open('admission_lr_model.pickle', 'rb'))

In [103]:
test1 = scaler.transform([[337.000000,118.0,4.0,4.5,4.5,9.65,1]])
model.predict(test1)

array([0.95976005])

### Evaluation of the model

In [105]:
    x_test

array([[ 1.57229650e+00,  1.45770580e+00,  7.78905651e-01,
         6.32315491e-01,  1.73062093e-02,  1.59474966e+00,
         8.86405260e-01],
       [-2.30667922e-01,  1.34360100e-01,  7.78905651e-01,
         1.13735981e+00,  5.58125251e-01,  7.67219636e-01,
         8.86405260e-01],
       [-1.40519701e-01, -3.61894539e-01, -9.94658860e-01,
        -1.38786180e+00, -1.06433187e+00, -1.53331383e+00,
        -1.12815215e+00],
       [ 5.12433309e-15,  2.99778313e-01, -1.07876604e-01,
        -3.77773152e-01, -5.23512832e-01,  1.87948619e-01,
        -1.12815215e+00],
       [ 8.51110730e-01,  7.96032952e-01, -1.07876604e-01,
         1.27271169e-01, -5.23512832e-01,  7.83770236e-01,
         8.86405260e-01],
       [ 1.12155539e+00,  6.30614739e-01,  7.78905651e-01,
         1.13735981e+00,  5.58125251e-01,  1.08168105e+00,
         8.86405260e-01],
       [-2.39422523e+00, -1.18898560e+00, -1.88144112e+00,
        -1.89290612e+00, -1.60515091e+00, -1.68226924e+00,
        -1.1281521

In [106]:
y_test

361    0.93
73     0.84
374    0.39
155    0.77
104    0.74
       ... 
220    0.76
176    0.90
320    0.75
153    0.79
231    0.74
Name: Chance_of_Admit, Length: 125, dtype: float64

In [107]:
# Calculating r-squared
lr.score(x_test, y_test)

0.8175497115836482

In [110]:
# Adjusted r-squared
def adj_r2(x,y):
    r2 = lr.score(x,y)
    n = x.shape[0]
    p = x.shape[1]
    adjusted_r2 = 1-(1-r2)*(n-1)/(n-p-1)
    return adjusted_r2

In [111]:
adj_r2(x_test, y_test)

0.806633882362157

In [112]:
# Checking the equestion
lr.coef_, lr.intercept_

(array([0.022218  , 0.01962107, 0.00252948, 0.00128856, 0.01538998,
        0.07278466, 0.01299267]),
 0.722535407687509)

 ### Regularization 

### Lasso

In [125]:
lasso_cv = LassoCV(cv=5, max_iter=200000000, normalize=True)
lasso_cv.fit(x_train, y_train)

LassoCV(cv=5, max_iter=200000000, normalize=True)

In [126]:
lasso_cv.alpha_

5.9937188109793075e-05

In [127]:
lasso = Lasso(alpha=lasso_cv.alpha_)
lasso.fit(x_train, y_train)

Lasso(alpha=5.9937188109793075e-05)

In [128]:
lasso.score(x_test, y_test)

0.8175868503894765

#### Ridge

In [136]:
alphas = np.random.uniform(0,10, 50)

In [137]:
ridge_cv = RidgeCV(alphas=alphas, cv=10, normalize=True)

In [138]:
ridge_cv.fit(x_train, y_train)

RidgeCV(alphas=array([6.8446173 , 8.76546689, 7.77925322, 3.10112758, 8.55602146,
       1.14971714, 3.73872013, 4.05077772, 0.64217275, 6.32732219,
       7.20001539, 4.78995843, 6.79064201, 5.13327725, 2.46851418,
       3.46218572, 3.11047505, 4.66951192, 8.3615347 , 0.13631026,
       9.25703213, 5.09451726, 0.13345377, 1.00400151, 3.92375409,
       4.75750802, 4.87983014, 5.86233356, 2.13869056, 9.36391679,
       8.3670917 , 7.30990863, 5.29717728, 5.73701709, 8.64640351,
       5.70360135, 3.25033679, 7.83932718, 5.32480629, 1.07028537,
       9.70931003, 6.65128084, 1.45514103, 3.6834711 , 2.72894413,
       2.60677081, 2.10606011, 2.99227264, 6.68887837, 5.7757523 ]),
        cv=10, normalize=True)

In [139]:
ridge_cv.alpha_

0.13345377439308526

In [140]:
ridge = Ridge(alpha=ridge_cv.alpha_)

In [141]:
ridge.fit(x_train, y_train)

Ridge(alpha=0.13345377439308526)

In [142]:
ridge.score(x_test, y_test)

0.8175628128024137

#### ElasticNet

In [143]:
elastic = ElasticNetCV(alphas=None, cv=10)
elastic.fit(x_train, y_train)

ElasticNetCV(cv=10)

In [144]:
elastic.alpha_

0.001519283042086672

In [145]:
elastic.l1_ratio_

0.5

In [146]:
elastic_lr = ElasticNet(alpha=elastic.alpha_, l1_ratio=elastic.l1_ratio_)

In [147]:
elastic_lr.fit(x_train, y_train)

ElasticNet(alpha=0.001519283042086672)

In [148]:
elastic_lr.score(x_test, y_test)

0.8179767633151169